File: _curso/6_ddd_modelagem_tatica_e_patterns/src/domain/customer/entity/customer.ts
import Address from '../value_object/address';

export default class Customer {
    private _id: string;
    private _name: string;
    private _address!: Address;
    private _active: boolean = true;
    private _rewardPoints: number = 0;

    constructor(id: string, name: string) {
        this._id = id;
        this._name = name;
        this.validate();
    }

    get id(): string {
        return this._id;
    }

    get name(): string {
        return this._name;
    }

    get rewardPoints(): number {
        return this._rewardPoints;
    }

    set Address(address: Address) {
        this._address = address;
    }

    changeName(name: string): void {
        this._name = name;
        this.validate();
    }

    changeAddress(address: Address): void {
        this._address = address;
        this.validate();
    }

    activate(): void {
        if (this._address === undefined) {
            throw new Error("Address is mandatory to activate a customer");
        }

        this._active = true;
    }

    deactivate(): void {
        this._active = false;
    }

    isActive(): boolean {
        return this._active;
    }

    validate(): void {
        if (this._id.length === 0) {
            throw new Error("Id is required");
        }

        if (this._name.length === 0) {
            throw new Error("Name is required");
        }
    }

    addRewardPoints(points: number) {
        this._rewardPoints += points;
    }
}

File: _curso/6_ddd_modelagem_tatica_e_patterns/python/src/domain/customer/entity/customer.py
from dataclasses import dataclass, field
from typing import Optional
from domain.customer.value_object.address import Address

@dataclass()
class Customer:
    __id: str
    __name: str
    __address: Optional[Address] = None
    __active: Optional[bool] = True
    __rewardPoints: Optional[int] = 0

    def __post_init__(self):
        self.validate()

    @property
    def id(self) -> str:
        return self.__id

    @property
    def name(self) -> str:
        return self.__name

    @property
    def reward_points(self) -> int:
        return self.__rewardPoints

    @property
    def  address(self) -> Address:
        return self.__address

    def set_address(self, address: Address) -> None:
        self.__address = address

    def changeName(self, name: str) -> None:
        self.__name = name
        self.validate()

    def changeAddress(self, address: Address) -> None:
        self.__address = address
        self.validate()

    def activate(self) -> None:
        if not isinstance(self.__address, Address):
            raise Exception("Address is mandatory to activate a customer")

        self.__active = True

    def deactivate(self) -> None:
        self.__active = False

    def is_active(self) -> bool:
        return self.__active

    def validate(self) -> None:
        if len(self.__id) == 0:
            raise Exception("Id is required")

        if len(self.__name) == 0:
            raise Exception("Name is required")

    def addRewardPoints(self, points: int) -> None:
        self.__rewardPoints += points